code/__HELPERS/unsorted.dm

```diff
@@ -1501,6 +1501,17 @@ GLOBAL_DATUM_INIT(dview_mob, /mob/dview, new)
 
 #define UNTIL(X) while(!(X)) stoplag()
 
+#define UNTIL_OR_TIMEOUT(X, __time) \
+	do {\
+		var/__start_time = world.time;\
+		while(!(X)) {;\
+			stoplag();\
+			if(__start_time + __time <= world.time) {;\
+				CRASH("UNTIL_OR_TIMEOUT hit timeout limit of [__time]");\
```

Comment:
A crash will end the proc call, and as a macro, that means that the proc that uses this macro will be the one ended. Are you sure you want a timeout to not fail gracefully? Consider `stack_trace` if your goal was only to have a verbose error.

Reply:
That's intended, yes. If the UNTIL_OR_TIMEOUT() of a `/datum/http_request` goes on until the specified `__time` timeout, then we know that the query won't work and should abort to prevent any further lag to the server.